src/jvmMain/kotlin/com/sdercolin/vlabeler/video/FindVideo.kt

```diff
@@ -0,0 +1,28 @@
+package com.sdercolin.vlabeler.video
+
+import java.io.File
+import java.io.FileNotFoundException
+
+sealed interface VideoFinder {
+    fun find(path: String, acceptableExtensions: List<String>): String
+}
+
+enum class FindVideoStrategy : VideoFinder {
+    SamePlaceOfReferenceAudio {
+        override fun find(path: String, acceptableExtensions: List<String>): String {
+            val fileName = path.substringBeforeLast('.')
+            return acceptableExtensions.asSequence().map { fileName + it }.filter { File(it).isFile }.firstOrNull() ?: (
+                throw FileNotFoundException(
```

Comment:
We should handle this error instead of just throwing it.

I suggest using `kotlin.Result` here (you can find many examples in the existing code):

```kotlin
override fun find(path: String, acceptableExtensions: List<String>): Result<String> {
    val fileName = path.substringBeforeLast('.')
    val existingFilePath =
        acceptableExtensions.asSequence().map { fileName + it }.filter { File(it).isFile }.firstOrNull()
    return if (existingFilePath != null) {
        Result.success(existingFilePath)
    } else {
        Result.failure(
            FileNotFoundException(
                "Video not found at the same place of audio ($path)" +
                    "with name $fileName and extensions among $acceptableExtensions",
            ),
        )
    }
}
```

On the caller side:
```kotlin
// VideoState.kt
suspend fun locatePath(currentSampleFile: File) {
    FindVideoStrategy.SamePlaceOfReferenceAudio.find(
        currentSampleFile.absolutePath,
        listOf(".mp4"),
    )
        .onSuccess { videoPath = it }
        .onFailure {
            if (it is FileNotFoundException) {
                // `AppState.snackbarState` can be passed via constructor 
                snackbarState.showSnackbar("PLEASE CREATE A LOCALIZED STRING") 
            } else {
                Log.error(it)
            }
        }
}

// App.kt
appState.project?.currentSampleFile?.let { currentSampleFile ->
    LaunchedEffect(currentSampleFile) {
        // I changed `Unit` to `currentSampleFile` here because I think every time when `currentSampleFile` changes, `locatePath` should be called again?
        appState.videoState.locatePath(currentSampleFile)
    }
    if (appState.isShowingVideo) {
        Video(appState.videoState, appState.playerState)
    }
}
```

Reply:
Please test if the error case works:
- No crash
- Readable and fixable: User can see a snackbar (toast) telling them there is no available video file (so that user can conduct a rename action or something else to fix)
- Can recover from error: When switching to another sample file which has corresponding video file, it can successfully locate the video and work as usual

Reply:
Tried that, but maybe we should postpone the error message when user opens the video.

Reply:
I agree. Maybe just move this part inside `Video(...)` so that it's only called when `appState.isShowingVideo` is true.

You may then need to wrap some code inside `Video()` with `if(videoState.videoPath!=null)` to avoid showing empty video UI or throwing exception.

```kotlin
// App.kt
if (appState.isShowingVideo) {
    Video(appState.videoState, appState.playerState, appState) // the last `appState` is passed as `ProjectStore`
}

// Video.kt
@Composable
fun Video(
    videoState: VideoState,
    playerState: PlayerState,
    projectStore: ProjectStore,
) {
    val currentSampleFile = projectStore.project?.currentSampleFile
    LaunchedEffect(currentSampleFile) {
        if (currentSampleFile != null) {
            videoState.locatePath(currentSampleFile)
        }
    }
    if (videoState.hasVideoPath) {
        // the UI code and other code that need the path
        ..
    }
    ..
```

Reply:
Done.